plugins/infinispan60/src/main/java/org/radargun/service/Infinispan60ServerTopologyHistory.java

```diff
@@ -80,4 +88,155 @@ public void afterStop(boolean graceful) {
+         if (topologyChangesInProgress == 0) {
+            topologyChangeMap.put(ALL_CACHES, false);
+         } else {
+            topologyChangeMap.put(ALL_CACHES, true);
+         }
+      } catch (Exception e) {
+         log.error("Failed to retrieve data from JMX", e);
+      }
+      return topologyChangeMap;
+   }
+
+   /**
+    * Uses the <code>joinComplete</code> and <code>stateTransferInProgress</code> attributes of the
+    * StateTransferManager component of each defined cache to determine if a rehash is in progress
+    * on this cache. If either attribute is <code>true</code>, a rehash is in progress for a
+    * particular cache.
+    * 
+    * @return a Map where the key is the cache name, and the value is <code>true</code> if a rehash
+    *         is in progress on the cache. Also includes an entry for {@link #ALL_CACHES} in case
+    *         any rehash are in progress
+    */
+   protected Map<String, Boolean> rehashInProcess() {
+      Map<String, Boolean> rehashMap = new HashMap<>();
+      int rehashesInProgress = 0;
+      try {
+         MBeanServerConnection connection = service.connection;
+         if (connection == null)
+            return rehashMap;
+         if (stateTransferManagerNames.isEmpty()) {
+            try {
+               stateTransferManagerNames = connection.queryNames(
+                     new ObjectName(String.format(JMX_CACHE_COMPONENT, service.jmxDomain, JMX_STATE_TRANSFER_MANAGER)),
+                     null);
+               rehashCacheNameMap = new HashMap<>();
+               for (ObjectName managerName : stateTransferManagerNames) {
+                  // Parse out cache name from ObjectName
+                  String fullCacheName = ObjectName.unquote(managerName.getKeyProperty("name"));
+                  String cacheName = fullCacheName.substring(0, fullCacheName.indexOf('('));
+                  rehashCacheNameMap.put(managerName.toString(), cacheName);
+               }
+            } catch (IOException e) {
+               // Server isn't started yet?
+               log.debug("Couldn't query StateTransferManager object names.", e);
+            }
+         }
+         for (ObjectName managerName : stateTransferManagerNames) {
+            List<Attribute> result = connection.getAttributes(managerName, JMX_REHASH_ATTRS).asList();
+            if (result.get(0).getValue().equals(false) || result.get(1).getValue().equals(true)) {
+               log.info(String.format("Rehash in progress. %s = %s and %s = %s", JMX_JOIN_COMPLETE_ATTR, result.get(0)
+                     .getValue(), JMX_STATE_TRANSFER_IN_PROGRESS_ATTR, result.get(1).getValue()));
+               rehashMap.put(rehashCacheNameMap.get(managerName.toString()), true);
+               rehashesInProgress++;
+            } else {
+               log.debug("No rehash in progress");
+               rehashMap.put(rehashCacheNameMap.get(managerName.toString()), false);
+            }
```

Comment:
Looks like topologyChangeInProcess & rehashInProcess contain some amount of duplicate code. Maybe you could find a way how to refactor it? (e.g. introducing a common template and plugging in specific behavior via dedicated object)